source/common/router/router.cc

```diff
@@ -1198,8 +1245,13 @@ void Filter::UpstreamRequest::onPoolReady(Http::StreamEncoder& request_encoder,
   }
 
   upstream_timing_.onFirstUpstreamTxByteSent(parent_.callbacks_->dispatcher().timeSource());
+
+  bool end_stream = !buffered_request_body_ && encode_complete_ && !encode_trailers_;
+  // If end_stream is set in headers, and there are metadata to send, delays end_stream. The case
+  // only happens when decoding headers filters return ContinueAndEndStream.
```

Comment:
Ah, I think this answers my question up above on how we handle the headers only condition.  I think we should probably handle it in the HCM so that other filters don't have to all one-off this.

Reply:
There are two cases where an empty data frame is needed: 1. the request is headers only request. 2. a filter decides to ignore data and trailers even if the request has data or trailers, aka the decoding_headers_only_  case. Case 1 can be easily handled in HCM. But case 2 is not easy to deal with in HCM. None of the methods I came up with is as simple as the current solution. I added tests for both cases in the latest version.

The purpose of decoding_headers_only_ filter is to skip passing data to the following filters. But if a filter inserts metadata in decodeHeaders() before or after decoding_headers_only_ filter, to deal with it in HCM, we need to pass the empty data frame through the following filters, which conflicts what decoding_headers_only_ filter is doing. So we need to add conditions to make sure only empty frames pass through the following filters, not the data from request, and it only passes through once. Another problem I encountered is which filter the empty data frame should start iterating from. Sometimes, decodeHeaders() returns stop in a filter after decoding_headers_only_ filter. If the continue_data_entry([here](https://github.com/envoyproxy/envoy/blob/5867e33124deac887c7100f9c3b3480e1fd67f42/source/common/http/conn_manager_impl.cc#L837)) is set to decoding_headers_only_ filter (it should be, because that's when we know if we need the empty data frame or not), commonContinue() will call the decodeHeaders in the stopped filter again without reset the iteration state. Since we don't need to pass the empty frame through filters, the safest way is to iterate the empty frame from the router filter (the last one), which I think may be equivalent to the current solution. I am open to any ideas or suggestions.

Reply:
I think it's possible this comment is confusing me.

If any intermediate filter added metadata, end stream would not have been set along with the headers by the time the content got to the router filter (or I think we have a problem).  So if just-before-router filter added metadata, the router filter would get headers with end stream false.

The way I see it what this code is doing is determining if this is a headers only request to send to upstream.  If there encode is complete but there's body, trailers, or metadata, it's not a headers only stream.

If it is the case that ContinueAndEndStream results in headers-with-end-stream-true going through the filter chain, wouldn't every filter up to the router filter be assuming it's done before it gets metadata?

Reply:
When the request is a header-only request, added metadata will unset end stream in headers, and router filter receives headers without end stream set. But if it's a ContinueAndEndStream case, and decodeHeaders() added metadata, I keep the end stream in headers to tell the following filters that request's data and trailers should not be parsed. In this case, metadata are guaranteed to arrive before headers at the router filter. Because addDecodedMetadata() calls decodeMetadata() in every filters before return. And decodeMetadata() can only return continue. The router filter will always get metadata before headers if decodeHeaders() adds metadata. Does that make sense?

If we unset end stream in ContinueAndEndStream case, I think it would be hard/messy to let the following filters skip parsing data and trailers.

Reply:
"The router filter will always get metadata before headers if decodeHeaders() adds metadata."

Is this still true with the decode_headers_called_ check in the HCM?  I thought that made sure that metadata never races before headers?

Reply:
Ah, you are right! With decode_headers_called_, metadata should always arrive after headers! decode_headers_called_ was created to handle a very rare case where a filter inserts new metadata, and another filter follows it returns StopAllIteration. To prevent metadata passing through a filter where the filter's header callback hasn't been called yet, decode_headers_called_ is used to indicate if header callback has been called. I only thought its use in StopAll cases, and didn't thought it has the impact of delay metadata for all statuses. Good catch!

It would simplify router filter's code then. We don't need to consider upstream request creation in decodeMetadata()!

But back to the main point in this comment thread, I think it's still possible that router filter receives headers with end stream set, and then it receives metadata because end_stream is used to tell the following filters that data and trailers should not be parsed.